core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,6 +539,107 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  /*
+   * For each pair of partition and log directory specified in the map, record the pair in the memory so that the partition
+   * will be created in the specified log directory when broker receives LeaderAndIsrRequest for the partition later.
+   *
+   * This API is currently only useful if the replica has not been created yet. We will be able to move replicas
+   * that are already created to the user-specified log directory after KIP-113 is fully implemented
+   *
+   */
+  def alterReplicaDir(partitionDirs: Map[TopicPartition, String]): Map[TopicPartition, Errors] = {
+    partitionDirs.map { case (topicPartition, destinationDir) =>
+      try {
+        if (!logManager.isLogDirOnline(destinationDir))
+          throw new KafkaStorageException(s"Log directory $destinationDir is offline")
+
+        // If the log for this partition has not been created yet:
+        // 1) Respond with ReplicaNotAvailableException for this partition in the AlterReplicaDirResponse
```

Comment:
I know this is in the KIP, but it still seems a little weird to throw exception in this case as the request is kink of legitimate. Not sure if there is a better way though.